pysrc/utils.py

```diff
@@ -68,3 +82,189 @@ def is_string(obj):
+        deps_manager: DepsManager class object (PIP, PIPENV, SETUPTOOLS).
+    Returns:
+        bool: True if `[` or `]` found in a package, False otherwise
+    """
+    for dependency in root_dependencies:
+        if '[' in dependency.name and ']' in dependency.name:
+            return True
+
+        elif (deps_manager is DepsManager.SETUPTOOLS
+              and '[' in dependency.line and ']' in dependency.line):
+            return True
+
+    return False
+
+
+def extract_dependencies(dependency_dict):
+    dependencies = []
+
+    def extract_names(dependency_dict):
+        for _, value in dependency_dict.items():
+            if isinstance(value, dict):
+                dependencies.append(value.get('name'))
+                if 'dependencies' in value:
+                    extract_names(value.get('dependencies'))
+
+    if 'dependencies' in dependency_dict:
+        extract_names(dependency_dict.get('dependencies'))
+
+    return dependencies
+
+
+def construct_tree_map(index):
+    """Construct tree representation of the pkgs from the index.
+
+    The keys of the dict representing the tree will be strings and the values
+    will be list of ReqPackage objects.
+
+    Args:
+        index: dist index i.e. index of pkgs by their keys
+
+    Returns:
+        dict: tree of pkgs and their dependencies
+    """
+    return dict(
+        (p.key, [ReqPackage(r, index.get(r.key)) for r in p.requires()])
+        for p in index.values()
+    )
+
+
+def establish_optional_dependencies(
+        venv_deps,
+        required_deps,
+        dist_tree,
+        deps_manager
+):
+    """Establishes if the project has optional dependencies or not.
+
+    Builds the dependency tree in reverse. The dependencies that are left out
+    of the tree, will be considered optional dependencies/dependencies that
+    were introduced by arbitrary identifiers.
+    Args:
+        venv_deps: dependencies installed in the venv.
+        required_deps: dependencies mentioned in the requirements file.
+        dist_tree: distribution tree based on the venv_deps.
+        deps_manager: DepsManager class object (PIP, PIPENV, SETUPTOOLS)
+
+    Returns:
+        list[str]: if optional dependencies were found.
+        empty list: if no optional dependencies were found.
+    """
+    has_optional_dependencies = check_optional_dependencies(
+        required_deps, deps_manager
+    )
+
+    if not has_optional_dependencies:
+        return []
+
+    dependency_map = construct_tree_map(venv_deps)
+
+    # Identify the root nodes
+    root_candidates = set(dependency_map.keys())
+
+    # Establish root dependencies
+    for dependency in dependency_map.values():
+        # Check if dependency has other dependencies
+        if dependency:
+            for subdep in dependency:
+                # Check if sub-dependency is amongst all other dependencies
+                if subdep.key in root_candidates:
+                    root_candidates.remove(subdep.key)
```

Comment:
General clean code best practice is to "keep left" with as few indentations as possible:

```suggestion
    # Establish root dependencies
for dependency in dependency_map.values():
    # Check if dependency has other dependencies
    if not dependency:
        continue

    for subdep in dependency:
        # Check if sub-dependency is amongst all other dependencies
        if subdep.key not in root_candidates:
            continue
        root_candidates.remove(subdep.key)
```

Reply:
Great catch, fixed in [7ba21cb](https://github.com/snyk/snyk-python-plugin/pull/234/commits/7ba21cb823349197437a93ed4395fa8ebd6ffaef) 🙏